azure/Guidelines.md

```diff
@@ -0,0 +1,241 @@
+
+### URL structure
+
+All services **MUST** expose their service to developers via the following URL pattern:
+
+```
+https://<service>.azure.net/<unit-of-multi-tenancy>/<service-defined-root>
+```
+
+Where:
+
+* **service** - the name of the service such as "blobstore", "servicebus", "directory", or "management"
+* **service-defined-root** - the root of the service-specific path, such as "blobcontainer", "myqueue", etc.
+* **unit-of-multi-tenancy** - refers to a globally unique moniker that identifies a unique container in the Azure service that has the following properties:
+
+  * This container is the boundary of isolation between different tenants of the service.
+  * Quotas as set and enforced at the level of this container - but there will be different limits for different operations; and operations will be service specific.
+  * Resources in the service are attached to this container and are tied to this container in terms of lifecycle. For example someone signs up, they get this container. If they unsubscribe (or don’t pay their bills) then cleanup of this container occurs and the resources associated with this container are cleaned up. Cleanup follows a state machine – the container and the resources attached to it are deactivated first (and can be easily restored if required), and if no response for some period then deleted.
+  * It is the container for billing – which means the owner of this container sees one bill for the resource usage of all azure services under this container’s identifier.
+
+For Azure PaaS services like SQL Azure, Azure Storage, Caching, etc., the unit of multi-tenancy is the Azure subscription id, which is a GUID. This ensures consistent access using the same URL pattern, and identifier, across all these services.
+
+When services produce URLs in response headers or bodies, they **MUST** use a consistent form – either always a GUID for tenant identifier or always a single verified domain - regardless of the URL used to reach the resource.
+
+### Direct endpoint URLs
+
+In addition to the required format above, services **MAY** also choose to expose direct endpoint for performance or routing reasons.  The direct endpoint should be discoverable by clients, to ensure that developers are presented with a consistent pattern for accessing Azure services.
+
+The format of the root of the direct endpoint **MUST** be as follows:
+
+```
+https://<tenant-id>-<service-defined-root>.<service>.azure.net
+```
+
+1. A request is made to the default end point (GET or HEAD).  For example:
+
+   ```
+   GET https://blobstore.azure.net/contoso.com/account1/container1/blob2
+   ```
+
+2. That request is returned with the `Content-Location` header set to the direct endpoint.  See [RFC2557]:
+
+   ```
+   200 OK
+   Content-Location: https://contoso-dot-com-account1.blobstore.azure.net/container1/blob2
```

Comment:
Is the Content-Location header identifying the content in response payload, or is the user expected to de-reference the URL in the Content-Location to access the blob?
If it is the latter, then that is not consistent with how [Content-Location](https://tools.ietf.org/html/rfc7231#section-3.1.4.2) is defined.

Reply:
This is a copy of the existing API guidelines, so it's been in there (like this) for a while.  @johanste or @johngossman - care to comment here?  I'm happy to update to correct an error here.

Reply:
@darrelmiller I don't understand the question

Reply:
@darrelmiller, this is not an alternate redirect implementation. The response body should (in the case of GET) include the data as per the specification. The `Content-Location` can be used for subsequent requests to avoid bouncing off the common front door etc. At least that's how I've read the specification. Please note that I don't think I've ever actually gotten a question on this in an API review.

Reply:
@johanste Thanks.  I was curious because I had also never seen that pattern come up before.